cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -186,12 +198,16 @@ class ARROW_EXPORT FileSystem : public std::enable_shared_from_this<FileSystem>
   virtual Result<FileInfo> GetFileInfo(const std::string& path) = 0;
   /// Same, for many targets at once.
   virtual Result<FileInfoVector> GetFileInfo(const std::vector<std::string>& paths);
+  /// Get file for file selector
+  virtual Result<FileInfoVector> GetFileInfo(const FileSelector& select);
   /// Same, according to a selector.
   ///
   /// The selector's base directory will not be part of the results, even if
   /// it exists.
   /// If it doesn't exist, see `FileSelector::allow_not_found`.
-  virtual Result<FileInfoVector> GetFileInfo(const FileSelector& select) = 0;
+  virtual Result<FileInfoVector> GetFileInfoDir(const FileSelector& select) = 0;
```

Comment:
I think `GetFileInfo(const FileSelector& select)` is sufficient.  If we want `GetFileInfoDir` and `GetFileInfoGlob` behind the scenes then that is fine but they should be private methods (or they could be protected methods if you're trying to provide a default implementation).  Otherwise I could call `GetFileInfoGlob` with a file selector with the directory type and it's not clear what that should do.

Reply:
I make `GetFileInfoDir` and `GetFileInfoGlob` methods protected